libmscore/style.cpp

```diff
@@ -2474,18 +2526,217 @@ void MStyle::save(XmlWriter& xml, bool optimize)
             _chordList.write(xml);
             xml.etag();
             }
-      xml.tag("Spatium", value(Sid::spatium).toDouble() / DPMM);
+      xml.tag("Spatium", value(Sid::spatium).toDouble() / DPMM);                 // 3.01
+      xml.tag("spatium", rint(value(Sid::spatium).toDouble() / DPI_F * MSCX_F)); // 3.01+ last one wins at load time
       xml.etag();
       }
 
+//--------------------------------------------------------------------------
+//   initPageLayout - only applied to MScore::_baseStyle and _defaultStyle
+//                    must wait until setMscoreLocale() is called in main()
+//--------------------------------------------------------------------------
+
+void MStyle::initPageLayout()
+      { // by default, units are millimeters or inches, depending on locale
+        // default preferences are initialized prior to default styles in main()
+      QPageLayout::Unit     unit;
+      QPageSize::PageSizeId psid;
+      if (MScore::testMode) {
+            unit = QPageLayout::Millimeter;
+            psid = QPageSize::A4;
+            }
+      else {
+            unit = QPageLayout::Unit(MScore::unitsValue());
+            psid = MStyle::isMetric(unit) ? QPageSize::A4
+                                          : QPageSize::Letter;
+            }
+      _pageSize = QPageSize(psid);           
+      _pageOdd  = MPageLayout();
+      _pageOdd.setPageSize(_pageSize);
+      _pageOdd.setUnits(QPageLayout::Millimeter);
+      _pageOdd.setMargins(QMarginsF(10, 10, 10, 20)); // default margins in millimeters
+      _pageOdd.setUnits(unit);                        // set to user locale units
+      _pageOdd.setOrientation(QPageLayout::Portrait);
+      _pageEven = MPageLayout(_pageOdd);
+
+      set(Sid::pageUnits, int(unit));
+      set(Sid::pageSize,  int(psid));
+
+      fromPageLayout(); // sync the styles
+      }
+
+//------------------------------------------------------------------------------
+//   from & toPageLayout convert between QPageLayout/QPageSize and style values
+//------------------------------------------------------------------------------
+void MStyle::fromPageLayout()
+      {
+      if (!MScore::testMode) { ///!!!Travis workaround
+            QRectF    rect = _pageOdd.fullRect(QPageLayout::Inch);
+            QMarginsF marg = _pageOdd.margins (QPageLayout::Inch);
+
+            if (abs(value(Sid::pageWidth) .toDouble() - rect.width())  >= 0.01)
```

Comment:
Extract the precision value as separate variable. Extract a method which compares two values with specified precision.